From b275b200bc61de34c33ad1272db7714d987f701f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bart=C5=82omiej=20Piotrowski?= <b@bpiotrowski.pl>
Date: Fri, 11 Oct 2019 10:34:22 +0200
Subject: [PATCH 8/8] Reject appdata without defined component type

---
 libappstream-glib/as-app-validate.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/libappstream-glib/as-app-validate.c b/libappstream-glib/as-app-validate.c
index 316f5a2..2e325db 100644
--- a/libappstream-glib/as-app-validate.c
+++ b/libappstream-glib/as-app-validate.c
@@ -1322,12 +1322,15 @@ as_app_validate (AsApp *app, guint32 flags, GError **error)
 	if (!as_app_validate_setup_networking (helper, error))
 		return NULL;
 
-	/* invalid component type */
-	if (as_app_get_kind (app) == AS_APP_KIND_UNKNOWN) {
+	switch (as_app_get_kind (app)) {
+	case AS_APP_KIND_UNKNOWN:
+	case AS_APP_KIND_GENERIC:
 		ai_app_validate_add (helper,
-				     AS_PROBLEM_KIND_ATTRIBUTE_INVALID,
-				     "<component> has invalid type attribute");
+			AS_PROBLEM_KIND_ATTRIBUTE_INVALID,
+			"<component> has invalid type attribute");
 
+	default:
+		break;
 	}
 	as_app_validate_check_id (helper, as_app_get_id (app));
 
-- 
2.23.0

